lib/src/editor.dart

```diff
@@ -0,0 +1,499 @@
+import 'dart:collection' show UnmodifiableListView;
+
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart';
+
+import 'equality.dart';
+import 'list_mutations.dart';
+import 'map_mutations.dart';
+import 'source_edit.dart';
+import 'utils.dart';
+import 'wrap.dart';
+
+/// An interface for modififying [YAML][1] documents while preserving comments and
+/// whitespaces.
+///
+/// Most modification methods require the user to pass in an Iterable<Object> path that
+/// holds the keys/indices to navigate to the element. Key equality is performed via
+/// `package:yaml`'s [YamlMap]'s key equality.
+///
+/// For example,
+///
+/// ```yaml
+/// a: 1
+/// b: 2
+/// c:
+///   - 3
+///   - 4
+///   - {e: 5, f: [6, 7]}
+/// ```
+///
+/// To get to `7`, our path will be `['c', 2, 'f', 1]`. The path for the base object is the
+/// empty array `[]`. All modification methods will return a [ArgumentError] if the path
+/// provided is invalid. Note also that that the order of elements in the path is important,
+/// and it should be arranged in order of calling, with the first element being the first
+/// key or index to be called.
+///
+/// To dump the YAML after all the modifications have been completed, simply call [toString()].
+///
+/// [1]: https://yaml.org/
+@sealed
+abstract class YamlEditor {
+  /// Returns a list of [SourceEdit]s describing the modifications performed on the
+  /// YAML string thus far.
+  UnmodifiableListView<SourceEdit> get edits;
+
+  factory YamlEditor(String yaml) => YamlStringEditor(yaml);
+
+  /// Parses the document to return [YamlNode] currently present at [path]. If no [YamlNode]s exist
+  /// at [path], [parseAt] will return a [YamlNode]-wrapped [orElse] if it is defined, or throw an
+  /// [ArgumentError] otherwise.
+  ///
+  /// Note that this behavior deviates slightly from standard [Map] behavior, which returns `null`
+  /// instead if the key is not present.
+  YamlNode parseAt(Iterable<Object> path, {Object orElse});
+
+  /// Sets [value] in the [path].
+  ///
+  /// Note that [assign] provides a different result as compared to a [remove] followed by an
+  /// [insertIntoList], because it preserves comments at the same level.
+  ///
+  /// Throws an [ArgumentError] if path is invalid, and throws the usual Dart errors otherwise (e.g.
+  /// [RangeError] if an index is negative or longer than list length).
```

Comment:
hmm, this makes a good argument for why we should have a custom `PathError extends/implements ArgumentError`..

Note. generally speaking dart core libraries do not specify what errors they throw, they do specify what `Exception`s they throw. Usually docs just say "if <something> then ... throws" in cases where <something> causes an error.
If it causes an exception it instead specifies what exception is thrown.

We could simply write here: `Throws if [path] is invalid.`

The down side over-specifying is that we can't change which errors we throw in the future (perhaps throwing new errors). Perhaps we should instead document on the `PathError` class that it is thrown when an _invalid path_ is used.